metals/src/main/scala/scala/meta/internal/metals/BloopServers.scala

```diff
@@ -121,6 +125,207 @@ final class BloopServers(
     }
   }
 
+  private def writeJVMPropertiesToBloopGlobalJsonFile(
+      bloopGlobalJsonFilePath: AbsolutePath,
+      bloopCreatedByMetalsFilePath: AbsolutePath,
+      requestedBloopJvmProperties: List[String],
+      maybeJavaHome: Option[String]
+  ): Try[Unit] = Try {
+
+    val javaOptionsString =
+      s"\"javaOptions\": [${requestedBloopJvmProperties.map(property => s"\"$property\"").mkString(", ")}]"
+
+    val jvmPropertiesString = maybeJavaHome
+      .map { javaHome =>
+        s"""|{
+            | $javaOptionsString,
+            | \"javaHome\": \"$javaHome\"
+            |}""".stripMargin
+      }
+      .getOrElse(s"{$javaOptionsString}")
+    bloopGlobalJsonFilePath.writeText(jvmPropertiesString)
+    bloopCreatedByMetalsFilePath.writeText(
+      bloopGlobalJsonFilePath.toNIO.toFile.lastModified().toString
+    )
+  }
+
+  private def getBloopGlobalJsonLastModifiedByMetalsTime(
+      bloopCreatedByMetalsFilePath: AbsolutePath
+  ): Long = Try {
+    bloopCreatedByMetalsFilePath.readText.toLong
+  }.getOrElse(0)
+
+  private def processUserPreferenceForBloopJvmProperties(
+      messageActionItem: MessageActionItem,
+      bloopGlobalJsonFilePath: AbsolutePath,
+      bloopCreatedByMetalsFilePath: AbsolutePath,
+      requestedBloopJvmProperties: List[String],
+      maybeJavaHome: Option[String],
+      reconnect: () => Future[BuildChange]
+  ): Future[Unit] = {
+    messageActionItem match {
+
+      case item
+          if item == Messages.BloopGlobalJsonFilePremodified.applyAndRestart =>
+        writeJVMPropertiesToBloopGlobalJsonFile(
+          bloopGlobalJsonFilePath,
+          bloopCreatedByMetalsFilePath,
+          requestedBloopJvmProperties,
+          maybeJavaHome
+        ) match {
+          case Failure(exception) => Future.failed(exception)
+          case Success(_) =>
+            shutdownServer()
+            reconnect().ignoreValue
+        }
+
+      case item
+          if item == Messages.BloopGlobalJsonFilePremodified.openGlobalJsonFile =>
+        val position = new Position(0, 0)
+        val range = new org.eclipse.lsp4j.Range(position, position)
+        val command = ClientCommands.GotoLocation.toExecuteCommandParams(
+          ClientCommands.WindowLocation(
+            bloopGlobalJsonFilePath.toURI.toString,
+            range
+          )
+        )
+        Future.successful(languageClient.metalsExecuteClientCommand(command))
+
+      case item
+          if item == Messages.BloopGlobalJsonFilePremodified.useGlobalFile =>
+        Future.successful(())
+    }
+  }
+
+  private def updateBloopGlobalJsonFileThenRestart(
+      bloopGlobalJsonFilePath: AbsolutePath,
+      bloopCreatedByMetalsFilePath: AbsolutePath,
+      requestedBloopJvmProperties: List[String],
+      maybeJavaHome: Option[String],
```

Comment:
I wonder also if we should update when just java home is set in case Metals and Bloop runs a different Java version. This could make the logic a bit more complicated, but I would treat `javaHome` as kind of additional property.

I've seen a bunch of issues related to that recently. We can follow up in a different PR for sure once this is merge or we can iterate on this one.

Opinions?

Reply:
I would do it in another PR.

Reply:
Ok cool! Let's do it in a different PR